sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/MessageFlux.java

```diff
@@ -0,0 +1,858 @@
+                        downstream.onComplete();
+                        return true;
+                    }
+                } else {
+                    Operators.onDiscard(messageDropped, downstream.currentContext());
+                    upstream.cancel();
+                    mediatorHolder.freeze();
+                    downstream.onComplete();
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        /**
+         * CONTRACT: Never invoke from the outside of serialized drain-loop.
+         * <br/>
+         * Request the next mediator if the operator is not in a termination signaled state and error is
+         * retriable and the retry is not exhausted. If there is a non-retriable or retry exhaust error,
+         * then proceed with error-ed termination of the operator.
+         *
+         * @param error the error that leads to error-ed termination of the last mediator or {@code null}
+         *              if terminated with completion.
+         * @param mediatorHolder the mediator holder.
+         */
+        private void tryRequestNextMediator(Throwable error, MediatorHolder mediatorHolder) {
+            if (cancelled || done) {
+                // To terminate the operator, the downstream signaled cancellation or upstream signaled error
+                // or completion. The next drain-loop iteration as a result of that signalling terminate
+                // the operator through one of the 'terminateIf*' methods followed by placing tombstone on
+                // the drain-loop.
+
+                // Below flag indicates there is no mediator currently, and there won't be one in the future,
+                // which shows the current mediator is flushed. When 'completeAfterMediatorFlush' is set
+                // and upstream signals operator completion, the flag 'noNextMediator' tells the next drain-loop
+                // iteration that it can terminate downstream gracefully.
+                mediatorHolder.noNextMediator = true;
+                return;
+            }
+
+            final Duration delay;
+            if (error == null) {
+                // TODO (anu): decide the back-off duration when no error.
+                delay = Duration.ofSeconds(1);
+            } else {
+                // error != null
+                // TODO (anu): it seems Event Hubs Processor retries only on receiver completion, not when error-ed. If true, we need a flag.
+                final int attempt = retryAttempts.incrementAndGet();
+                delay = retryPolicy.calculateRetryDelay(error, attempt);
+                if (delay == null) {
+                    // TODO (anu): log non-retriable or retry exhausted error
+                    // Invoking 'onError' to signal the next drain-loop iteration to terminate the operator
+                    // with 'error',
+                    onError(error);
```

Comment:
Decide if inspecting the error code LINK_STOLEN for link which is not current is required.